symFileManager.py

```diff
@@ -28,6 +30,95 @@ def Lookup(self, address):
   def GetEntryCount(self):
     return self.entryCount
 
+# Manage in disk symbol cache
+class SymDiskCache(object):
+  def __init__(self, cachePath, diskQuota):
+    mkdir_p(cachePath)
+    self.cachePath = cachePath
+    self.diskQuota = diskQuota
+
+    self.usedQuota, fileList = self._walk(cachePath)
+
+    # sort the symbol relative to the last access time
+    fileList.sort(cmp=lambda a,b: cmp(a[1], b[1]))
+
+    # filter only the paths
+    self.lru = list(map(lambda x: x[0], fileList))
+
+    self.checkQuota()
+
+  def fetch(self, pathSuffix):
+    r"fetch a symbol file from the cache"
+    path = self.absPath(pathSuffix)
+    symbolInfo = None
+    if os.path.exists(path):
+      with open(path, 'r') as f:
+        symbolInfo = pickle.load(f)
+    return symbolInfo
+
+  def store(self, symbolInfo, pathSuffix):
+    r"""Store a symbol file in the cache.
+
+    The caller must call use(pathSuffix) before this
+    method.
+    """
+    path = self.absPath(pathSuffix)
+    dirname = os.path.dirname(path)
+    mkdir_p(dirname)
+
+    with open(path, 'w') as f:
+      pickle.dump(symbolInfo, f, pickle.HIGHEST_PROTOCOL)
+
+    # We increased the used cache size, ensure we did not
+    # crosses the size limit
+    self.usedQuota += os.path.getsize(path)
+
+  def use(self, pathSuffix):
+    r"Mark the cached sym file as recently used."
+    normPath = self.normPath(pathSuffix)
+    path = self.absPath(normPath)
+    if os.path.exists(path):
+      try:
+        self.lru.remove(normPath)
+      except ValueError:
+        pass
+      self.lru.append(normPath)
+
+  def checkQuota(self):
+    r"Make sure we did not crosses the quota limit."
+    while self.usedQuota > self.diskQuota and self.lru:
+      # The least recently used symbol is the first item in the list
+      path = self.lru[0]
+      del self.lru[0]
+
+      # Remove from the disk
+      path = self.absPath(path)
+      fileSize = os.path.getsize(path)
+      os.remove(path)
+      self.usedQuota -= fileSize
+
+  def absPath(self, pathSuffix):
+    r"Get the absolute path of the symbol file."
+    return os.path.join(self.cachePath, pathSuffix)
+
+  def normPath(self, pathSuffix):
+    return os.path.normpath(pathSuffix)
+
+  def _walk(self, path):
+    r"Walk through the cache directory collecting files."
+    totalSize = 0
+    fileList = []
+
+    for root, _, filenames in os.walk(path):
+      for filename in filenames:
+        filePath = os.path.join(root, filename)
+        stat = os.stat(filePath)
+        totalSize += stat.st_size
+        relativePath = os.path.relpath(filePath, self.cachePath)
```

Comment:
`filename` is relative to `root`, `relativePath` is relative to `self.cachePath`.